my_enumerables.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+# rubocop : disable Metrics/ModuleLength
+# rubocop : disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+# rubocop : disable Metrics/MethodLength
+# rubocop : disable Style/Documentation
+# rubocop : disable Metrics/AbcSize
+
+module Enumerable
+  def my_each
+    if is_a?(Array)
+      0.upto size - 1 do |val|
+        yield(self[val]) if block_given?
+      end
+    elsif is_a?(Range)
+      each do |val|
+        yield(val) if block_given?
+      end
+    else
+      hash_keys = keys
+      0.upto hash_keys.length - 1 do |num|
+        key = hash_keys[num]
+        yield(key, self[key])
+      end
+    end
+  end
+
+  def my_each_with_index
+    0.upto size - 1 do |val|
+      yield(self[val], val)
+    end
+  end
+
+  def my_select
+    return_arr = []
+    return_hash = {}
+    if is_a?(Hash)
+      my_each do |key, val|
+        return_hash[key] = val if yield(key, val)
+      end
+      return_hash
+    else
+      my_each do |val|
+        return_arr.push(val) if yield(val)
+      end
+      return_arr
+    end
+  end
+
+  def my_all?
+    ans = false
+    if is_a?(Hash)
+      my_each do |key, val|
+        if yield(key, val)
+          ans = true
+        else
+          ans = false
+          break
+        end
+      end
+    else
+      my_each do |val|
+        if yield(val)
+          ans = true
+        else
+          ans = false
+          break
+        end
+      end
+    end
+    ans
+  end
+
+  def my_any?
+    ans = false
+    if is_a?(Hash)
+      my_each do |key, val|
+        if yield(key, val)
+          ans = true
+          break
+        else
+          ans = false
+        end
+      end
+    else
+      my_each do |val|
+        if yield(val)
+          ans = true
+          break
+        else
+          ans = false
+        end
+      end
+    end
+    ans
+  end
+
+  def my_none?
+    ans = false
+    if is_a?(Hash)
+      my_each do |key, val|
+        if yield(key, val)
+          ans = false
+          break
+        else
+          ans = true
+        end
+      end
+    else
+      my_each do |val|
+        if yield(val)
+          ans = false
+          break
+        else
+          ans = true
+        end
+      end
+    end
+    ans
+  end
+
+  def my_count(args = nil)
+    ans = 0
+    if block_given?
+      my_each do |val|
+        ans += 1 if yield(val)
+      end
+    elsif args.nil?
+      ans = size
+    else
+      my_each do |val|
+        ans += 1 if val == args
+      end
+    end
+    ans
+  end
+
+  def my_map(args = nil)
+    return_arr = []
+    if block_given? && args.is_a?(Proc)
+      my_each do |val|
+        return_arr.push(args.call(val))
+      end
+    elsif !block_given? && args.is_a?(Proc)
+      my_each do |val|
+        return_arr.push(args.call(val))
+      end
+    else
+      my_each do |val|
+        return_arr.push(yield(val))
+      end
+    end
+    return_arr
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
    # LocalJumpError:
        no block given (yield)
      # ./my_enumerables.rb:150:in `block in my_map'
      # ./my_enumerables.rb:13:in `block in my_each'
      # ./my_enumerables.rb:12:in `upto'
      # ./my_enumerables.rb:12:in `my_each'
      # ./my_enumerables.rb:149:in `my_map'
  ```